src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs

```diff
@@ -402,9 +405,29 @@ matchNonVarToPattern tools first second
                 makeAndPredicate
                     predicate
                     (substitutionToPredicate substitution)
-        return
+        return . return $ -- MonadCounter m => m (Maybe a)
             PredicateSubstitution
                 { predicate = finalPredicate
                 , substitution = []
                 }
-matchNonVarToPattern _ _ _ = Nothing
+
+checkVariableEscape
+    :: ( MetaOrObject level
+        , Show (variable Object)
+        , Show (variable Meta)
+        , Ord (variable Object)
+        , Ord (variable Meta)
+        , Given (SymbolOrAliasSorts level)
+        , SortedVariable variable
+        , Eq (variable level)
+        , Ord (variable level)
+        , Show (variable level))
+    => [variable level]
+    -> PredicateSubstitution level variable
+    -> PredicateSubstitution level variable
+checkVariableEscape vars predSubst
+  | any (`Set.member` freeVars) vars = error
+        "quantified variables in substitution or predicate escaping context"
```

Comment:
I know that "escaping context" describes what happens, but it would be nice to have a more clear comment somewhere about why this is a bad thing (e.g. saying that this is called when processing a quantified pattern, with a substitution-predicate that would be moved above the quantifier, in which case they can't contain the quantified variable; or something similar).